Services/Cache.cs

```diff
@@ -55,89 +62,154 @@ public class Cache : ICache
+
+                    this.log.Warn($"Service IoTHubManager and/or Simulation are not ready. Retry after {serviceQueryInterval}", () => { });
+                    await Task.Delay(serviceQueryInterval);
                     continue;
                 }
 
-                if (cache != null)
+                // Lock the cache for writing
+                ValueApiModel model;
+                try
                 {
-                    CacheValue model = JsonConvert.DeserializeObject<CacheValue>(cache.Data);
-                    model.Rebuilding = true;
-                    var response = await this.storageClient.UpdateAsync(CACHE_COLLECTION_ID, CACHE_KEY, JsonConvert.SerializeObject(model), cache.ETag);
-                    etag = response.ETag;
+                    var cacheValue = cache == null ? new CacheValue() : JsonConvert.DeserializeObject<CacheValue>(cache.Data);
+                    cacheValue.Rebuilding = true;
+
+                    model = await this.storageClient.UpdateAsync(CACHE_COLLECTION_ID, CACHE_KEY, JsonConvert.SerializeObject(cacheValue), cache?.ETag);
                 }
-                else
+                catch (ConflictingResourceException)
                 {
-                    var response = await this.storageClient.UpdateAsync(CACHE_COLLECTION_ID, CACHE_KEY, JsonConvert.SerializeObject(new CacheValue { Rebuilding = true }), null);
-                    etag = response.ETag;
+                    this.log.Warn("Cache rebuilding failed due to conflict while writing lock. Retry soon", () => { });
+                    continue;
                 }
 
-                var value = JsonConvert.SerializeObject(new CacheValue
-                {
-                    Rebuilding = false,
-                    Tags = twinNames.Tags,
-                    Reported = reportedNames
-                });
-
+                // Write the cache to storage                
                 try
                 {
-                    await this.storageClient.UpdateAsync(CACHE_COLLECTION_ID, CACHE_KEY, value, etag);
-                    return;
+                    await this.storageClient.UpdateAsync(
+                        CACHE_COLLECTION_ID,
+                        CACHE_KEY,
+                        JsonConvert.SerializeObject(new CacheValue
+                        {
+                            Rebuilding = false,
+                            Tags = twinNames.Tags,
+                            Reported = twinNames.ReportedProperties
+                        }),
+                        model.ETag);
+
+                    this.log.Info("Cache rebuilding completed", () => { });
+                    return true;
                 }
                 catch (ConflictingResourceException)
                 {
-                    this.log.Info("rebuild Conflicted ", () => {});
-                    continue;
+                    this.log.Warn("Cache rebuilding failed due to conflict while writing value. Retry soon", () => { });
```

Comment:
sleep/delay before looping?

Reply:
This exception will only be raised if the ETag is not matched (document in CosmosDb was updated by some other client). It will not be a frequenting event. So sleep/delay will not decrease the probability of conflict. On the other hand, considering we applied a "write lock" pattern here, it will be better if we could complete the whole action and release the lock earlier.